src/lib/support/CHIPMemString.h

```diff
@@ -34,18 +36,69 @@ namespace Platform {
  *
  * @param[in]  string           String to be copied.
  *
- * @param[in]  length           Length of the string to be copied. If zero, `strlen(string)`
- *                              will be used. Like `strncpy()`, if the `string` is shorter
- *                              then the remaining space up to `length` will be filled with
- *                              null bytes.
+ * @param[in]  length           Length to be copied. If zero, `strlen(string)` will be used.
+ *                              Like `strncpy()`, if the `string` is shorter than `length`,
+ *                              then the remaining space will be filled with null bytes.
+ *                              Like `strndup()` but unlike `strncpy()`, the result is always
+ *                              null-terminated.
  *
  * @retval  Pointer to a null-terminated string in case of success.
  * @retval  `nullptr` if memory allocation fails.
  *
  */
 extern char * MemoryAllocString(const char * string, size_t length = 0);
 
+/**
+ * Copies a C-style string.
+ *
+ * This differs from `strncpy()` in some important ways:
+ *  - `dest` can be nullptr, in which case no copy is attempted, and the function returns nullptr.
+ *  - `source` can be nullptr, in which case the destination is an empty string.
+ *  - A non-nullptr result is always null-terminated.
+ * Like `strncpy()`, if the `string` is shorter than `length` then the remaining space will be
+ * filled with null bytes.
+ *
+ * @param[in]  dest             Destination string buffer (which must be at least `length`+1 bytes)
+ *                              or nullptr.
+ *
+ * @param[in]  source           String to be copied.
+ *
+ * @param[in]  length           Length to be copied.
+ *
+ * @retval  Same as `dest`.
+ */
+extern char * CopyString(char * dest, const char * source, size_t length);
+
+/**
+ * Represents a C string in a ScopedMemoryBuffer.
+ */
+
+class ScopedMemoryString : public ScopedMemoryBuffer<char>
+{
+public:
+    /**
+     * Create a ScopedMemoryString.
+     *
+     * @param[in]  string           String to be copied. If `nullptr`, treated as a zero-length string.
+     *
+     * @param[in]  length           Length to be copied. If zero, `strlen(string)` will be used.
+     *                              Like `strncpy()`, if the `string` is shorter than `length`,
+     *                              then the remaining space will be filled with null bytes.
+     *                              Like `strndup()` but unlike `strncpy()`, the result is always
+     *                              null-terminated.
+     */
+    ScopedMemoryString(const char * string, size_t length = 0)
+    {
+        if (string && !length)
+        {
+            length = strlen(string);
+        }
+        if (string || length)
```

Comment:
If string is NULL but length is non-zero, is there still a string to copy?

Reply:
The effect is as if the string is empty. I can remove this if preferred.